Pokémon Emerald 3Ds Dual Screen changes to pret/pokeemerald 76463dac15cad36aca5e2b3f6366abf9e53f814f: Black/White sprites
Applies after 0005-port-hooks.patch; apply with tools/bootstrap.py.

diff --git a/src/decompress.c b/src/decompress.c
--- a/src/decompress.c
+++ b/src/decompress.c
@@ -9,6 +9,14 @@ EWRAM_DATA ALIGNED(4) u8 gDecompressionBuffer[0x4000] = {0};
 
 static void DuplicateDeoxysTiles(void *pointer, s32 species);
 
+#ifdef PLATFORM_3DS
+// The BW sprites learn each Pokemon picture as it's decompressed (3ds_bw.h).
+void PortBw_PicLoaded(const struct CompressedSpriteSheet *sheet, const void *dest);
+#define BW_PIC_LOADED(sheet, dest) PortBw_PicLoaded(sheet, dest)
+#else
+#define BW_PIC_LOADED(sheet, dest)
+#endif
+
 void LZDecompressWram(const u32 *src, void *dest)
 {
     LZ77UnCompWram(src, dest);
@@ -68,6 +76,7 @@ void DecompressPicFromTable(const struct CompressedSpriteSheet *src, void *buffe
     else
         LZ77UnCompWram(src->data, buffer);
     DuplicateDeoxysTiles(buffer, species);
+    BW_PIC_LOADED(species > NUM_SPECIES ? &gMonFrontPicTable[0] : src, buffer);
 }
 
 void HandleLoadSpecialPokePic(const struct CompressedSpriteSheet *src, void *dest, s32 species, u32 personality)
@@ -84,6 +93,9 @@ void HandleLoadSpecialPokePic(const struct CompressedSpriteSheet *src, void *des
 
 void LoadSpecialPokePic(const struct CompressedSpriteSheet *src, void *dest, s32 species, u32 personality, bool8 isFrontPic)
 {
+#ifdef PLATFORM_3DS
+    const struct CompressedSpriteSheet *used = src;
+#endif
     if (species == SPECIES_UNOWN)
     {
         u16 i = GET_UNOWN_LETTER(personality);
@@ -98,10 +110,16 @@ void LoadSpecialPokePic(const struct CompressedSpriteSheet *src, void *dest, s32
             LZ77UnCompWram(gMonBackPicTable[i].data, dest);
         else
             LZ77UnCompWram(gMonFrontPicTable[i].data, dest);
+#ifdef PLATFORM_3DS
+        used = isFrontPic ? &gMonFrontPicTable[i] : &gMonBackPicTable[i];
+#endif
     }
     else if (species > NUM_SPECIES) // is species unknown? draw the ? icon
     {
         LZ77UnCompWram(gMonFrontPicTable[0].data, dest);
+#ifdef PLATFORM_3DS
+        used = &gMonFrontPicTable[0];
+#endif
     }
     else
     {
@@ -109,6 +127,7 @@ void LoadSpecialPokePic(const struct CompressedSpriteSheet *src, void *dest, s32
     }
 
     DuplicateDeoxysTiles(dest, species);
+    BW_PIC_LOADED(used, dest);
     DrawSpindaSpots(species, personality, dest, isFrontPic);
 }
 
@@ -319,10 +338,14 @@ void DecompressPicFromTable_2(const struct CompressedSpriteSheet *src, void *buf
     else
         LZ77UnCompWram(src->data, buffer);
     DuplicateDeoxysTiles(buffer, species);
+    BW_PIC_LOADED(species > NUM_SPECIES ? &gMonFrontPicTable[0] : src, buffer);
 }
 
 void LoadSpecialPokePic_2(const struct CompressedSpriteSheet *src, void *dest, s32 species, u32 personality, bool8 isFrontPic) // a copy of LoadSpecialPokePic
 {
+#ifdef PLATFORM_3DS
+    const struct CompressedSpriteSheet *used = src;
+#endif
     if (species == SPECIES_UNOWN)
     {
         u16 i = GET_UNOWN_LETTER(personality);
@@ -337,10 +360,16 @@ void LoadSpecialPokePic_2(const struct CompressedSpriteSheet *src, void *dest, s
             LZ77UnCompWram(gMonBackPicTable[i].data, dest);
         else
             LZ77UnCompWram(gMonFrontPicTable[i].data, dest);
+#ifdef PLATFORM_3DS
+        used = isFrontPic ? &gMonFrontPicTable[i] : &gMonBackPicTable[i];
+#endif
     }
     else if (species > NUM_SPECIES) // is species unknown? draw the ? icon
     {
         LZ77UnCompWram(gMonFrontPicTable[0].data, dest);
+#ifdef PLATFORM_3DS
+        used = &gMonFrontPicTable[0];
+#endif
     }
     else
     {
@@ -348,6 +377,7 @@ void LoadSpecialPokePic_2(const struct CompressedSpriteSheet *src, void *dest, s
     }
 
     DuplicateDeoxysTiles(dest, species);
+    BW_PIC_LOADED(used, dest);
     DrawSpindaSpots(species, personality, dest, isFrontPic);
 }
 
@@ -369,6 +399,7 @@ void DecompressPicFromTable_DontHandleDeoxys(const struct CompressedSpriteSheet
         LZ77UnCompWram(gMonFrontPicTable[0].data, buffer);
     else
         LZ77UnCompWram(src->data, buffer);
+    BW_PIC_LOADED(species > NUM_SPECIES ? &gMonFrontPicTable[0] : src, buffer);
 }
 
 void HandleLoadSpecialPokePic_DontHandleDeoxys(const struct CompressedSpriteSheet *src, void *dest, s32 species, u32 personality)
@@ -385,6 +416,9 @@ void HandleLoadSpecialPokePic_DontHandleDeoxys(const struct CompressedSpriteShee
 
 void LoadSpecialPokePic_DontHandleDeoxys(const struct CompressedSpriteSheet *src, void *dest, s32 species, u32 personality, bool8 isFrontPic)
 {
+#ifdef PLATFORM_3DS
+    const struct CompressedSpriteSheet *used = src;
+#endif
     if (species == SPECIES_UNOWN)
     {
         u16 i = GET_UNOWN_LETTER(personality);
@@ -399,16 +433,23 @@ void LoadSpecialPokePic_DontHandleDeoxys(const struct CompressedSpriteSheet *src
             LZ77UnCompWram(gMonBackPicTable[i].data, dest);
         else
             LZ77UnCompWram(gMonFrontPicTable[i].data, dest);
+#ifdef PLATFORM_3DS
+        used = isFrontPic ? &gMonFrontPicTable[i] : &gMonBackPicTable[i];
+#endif
     }
     else if (species > NUM_SPECIES) // is species unknown? draw the ? icon
     {
         LZ77UnCompWram(gMonFrontPicTable[0].data, dest);
+#ifdef PLATFORM_3DS
+        used = &gMonFrontPicTable[0];
+#endif
     }
     else
     {
         LZ77UnCompWram(src->data, dest);
     }
 
+    BW_PIC_LOADED(used, dest);
     DrawSpindaSpots(species, personality, dest, isFrontPic);
 }
 
